### PR TITLE
Extract Gutenboarding language picker into @automattic/language-picker

### PR DESCRIPTION
**Pull request by sarayourfriend**
#### Changes proposed in this Pull Request

* Extract the Gutenboarding language picker based on `@wordpress/components` into `@automattic/language-picker` allowing us to use it both in Gutenboarding and in Calypso.
* Replaces the content of the language picker in Gutenboarding with the extracted version.

#### Testing instructions

* Open `/new/language-modal` and confirm that it matches visually and behaviorally with production.

### CODE REVIEW

**Comment by simison** [comment]
Using this inside a modal from `@wordpress/components` might be interesting step as well BTW.

**Comment by saramarcondes** [comment]
@simison I love that idea! When we eventually implement the resulting component into the Calypso modal we can try to swap the `Dialog` for `Modal`

File: package.json
Comment (roo2) [inline package.json:157]: when I run this, I get an error on the storybook page in the browser "Exports is read only"

**Review by roo2 — COMMENTED** [review]
This looks good and I tested it with `/new?flags=gutenboarding/language-picker` turned on and it works well.

I just couldn't get the storybook working, can you please re-test that?

Also, did you consider adding this to `packages/components` rather than its own package? I think either way would be ok so no pressure to change it.

File: packages/language-picker/src/index.ts
Comment (tyxla) [inline packages/language-picker/src/index.ts:2]: Not sure what we're trying to do here, could you please elaborate?

File: packages/language-picker/src/language-picker.tsx
Comment (tyxla) [inline packages/language-picker/src/language-picker.tsx:97]: I see we're using `@wordpress/i18n` here. This relies on the locale data to be already set. AFAIK, `CalypsoI18nProvider` currently does that for both Calypso-style localizations and Gutenberg-style localizations. But using that in a package assumes that this is somehow handled already in the consumer of the package. What are your thoughts about this?

also cc @automattic/i18n about it

File: packages/language-picker/src/language-picker.tsx
Comment (tyxla) [inline packages/language-picker/src/language-picker.tsx:20]: Not sure why we're renaming `__()` to `__translate()` here. Perhaps we should be explicit that we're using `@wordpress/i18n`?

Comment (tyxla) [inline packages/language-picker/src/language-picker.tsx:20]: It feels a bit off to pass `translate` as an argument here. Did we mean to use it directly in the code? A function that returns a string should be enough IMHO.

**Review by tyxla — COMMENTED** [review]
This is exciting, thank you for the great work @saramarcondes!

I've left some comments, but this is a really great start! Let me know what you think.

File: packages/language-picker/src/language-picker.tsx
Comment (yuliyan) [reply packages/language-picker/src/language-picker.tsx:97]: If there's no provider, `@automattic/react-i18n` will use `@worpdress/i18n` instance with empty locale data, and therefore string won't get translated. A good proposal for `@automattic/react-i18n` might be to fall back to the default `@wordpress/i18n` instead of creating an empty one.

File: packages/language-picker/src/language-picker.tsx
Comment (saramarcondes) [reply packages/language-picker/src/language-picker.tsx:20]: This isn't renaming, it's the type annotation. i18n-calypso ultimately uses `@wordpress/i18n` does it not? `translate` is basically the same as the `I18n` function `__`, right?

Comment (saramarcondes) [reply packages/language-picker/src/language-picker.tsx:20]: The language groups are defined in a static file, not inside any specific context. I'm not sure how we could have translated names without passing translate to the static file unless we changed the way language groups are defined (as in, defining them in a specific runtime context rather than just a generic module). What are your thoughts for the alternatives to make this work with translated names still?

File: package.json
Comment (saramarcondes) [reply package.json:157]: I'm not sure why this suddenly isn't working 🤔 Reading the error below:

```./packages/language-picker/node_modules/@automattic/react-i18n/node_modules/@wordpress/is-shallow-equal/lib/objects.js/<@http://localhost:50609/vendors~main.74bd29d19c2e02c9b421.bundle.js:213500:1
```

It appears to be an issue within a dependency. The storybook worked before so I'm wondering if at some point I changed something while working on it that caused it to break (like updating the dependencies 🤔)

**Comment by saramarcondes** [comment]
@roo2 The reason I didn't put it into `@automattic/components` (I've updated the title) is because moving it in there was causing it to get imported on app-start (because those components are always used) and because this depends on the non-SSR compatible `@wordpress/components` it would cause the app to crash when starting. Having it in a separate package is a temporary workaround for that issue. Once the `@wordpress` scope is updated this PR can be rebased and then I could move it into the components (or we can merge this PR and then move it into components in a subsequent PR). The goal is that it would live in components though.

File: packages/language-picker/src/language-picker.tsx
Comment (tyxla) [reply packages/language-picker/src/language-picker.tsx:20]: I'd expect that we would just use a function that returns a string. See https://github.com/Automattic/wp-calypso/blob/try/extract-language-picker/client/lib/plans/plans-list.js#L25 for example.

By going with this approach, we:

* Ensure that localization functions will be invoked as late as possible, which is necessary to be sure we've loaded all the locale data already.
* Avoid the necessity to pass a specific localization function as an argument.
* Stay flexible and not enforce a specific i18n framework. By expecting just a function that returns a string, we let the package consumer decide what i18n tools to use, vs forcing them to use a particular one.

Let me know if that makes sense.

Comment (tyxla) [reply packages/language-picker/src/language-picker.tsx:20]: Forgive me if I've been unclear in my comment, but this is indeed renaming, even if it happens under the hood.

`i18n-calypso` actually does not use `@wordpress/i18n`. It is a custom library that was built before `@wordpress/i18n` and uses a different mechanism for handling localization and translations. This is separate from the fact that the `CalypsoI18nProvider` provides locale data for both `@wordpress/i18n` and `i18n-calypso`.

So actually, `__()` is the `@wordpress/i18n`, but `translate()` is the `i18n-calypso` function. Both are localization functions, and for simple translations they work the same way, but they still have a lot of functional differences. For example `translate()` supports some additional features (for example arguments, components, context), while for `@wordpress/i18n` you need to use different functions `sprintf()` or  `_x()` for example. 

And by expecting `__()` but naming the argument `translate`, we're essentially confusing the package user as to what kind of argument to provide and how it's being used. It's also possible to introduce bugs because of that behavior.

Finally, I believe that the actual translation mechanism shouldn't be handled (or enforced) by the package. I've talked about this a bit more in https://github.com/Automattic/wp-calypso/pull/46020/files#r504487702.

File: packages/language-picker/src/language-picker.tsx
Comment (tyxla) [reply packages/language-picker/src/language-picker.tsx:97]: Indeed, I've also explained more about that problem in https://github.com/Automattic/wp-calypso/pull/46020#discussion_r504487702

File: packages/language-picker/src/language-picker.stories.tsx
Comment (tyxla) [inline packages/language-picker/src/language-picker.stories.tsx:52]: How is this `_default` export expected to be used?

Comment (saramarcondes) [reply packages/language-picker/src/language-picker.stories.tsx:52]: This creates a `Default` example in Storybook. This is the same pattern followed by Gutenberg's usage of Storybook: https://github.com/WordPress/gutenberg/blob/master/packages/components/src/form-toggle/stories/index.js#L26

File: packages/language-picker/src/language-picker.tsx
Comment (saramarcondes) [reply packages/language-picker/src/language-picker.tsx:20]: Oh gosh, I misremembered. The original language picker uses `@automattic/react-i18n` which _does_ use `@wordpress/i18n`, which was why I chose that type.

I understand the issue now is that we have two separate functions, `translate` and `__`. The solution in your other comment makes perfect sense to me. Thank you for your patience!

Comment (saramarcondes) [reply packages/language-picker/src/language-picker.tsx:20]: Yes, thank you for the explanation and example!

File: packages/language-picker/package.json
Comment (scinos) [inline packages/language-picker/package.json:31]: We should import using the version

File: packages/language-picker/src/language-picker.tsx
Comment (saramarcondes) [reply packages/language-picker/src/language-picker.tsx]: @tyxla Does this also apply to localization that happens within the component itself (as in the case of the Regions label?) Would we just expect the consumer to dependency inject an i18n provider that had the signature `string => string`?

File: packages/language-picker/src/language-picker.tsx
Comment (tyxla) [inline packages/language-picker/src/language-picker.tsx:46]: There's a typo here: `defaultLananguageGroupId` -> `defaultLanguageGroupId`.

```suggestion
	defaultLanguageGroupId: string;
```

File: packages/language-picker/src/language-picker.tsx
Comment (tyxla) [inline packages/language-picker/src/language-picker.tsx:53]:
```suggestion
	defaultLanguageGroupId,
```

File: client/landing/gutenboarding/onboarding-block/language/index.tsx
Comment (tyxla) [inline client/landing/gutenboarding/onboarding-block/language/index.tsx:46]:
```suggestion
						defaultLanguageGroupId={ DEFAULT_LANGUAGE_GROUP }
```

File: packages/language-picker/src/language-picker.tsx
Comment (tyxla) [inline packages/language-picker/src/language-picker.tsx:53]: Additionally, I wonder if we can't always just default to the first item in `languageGroups`. That's not only one prop less, it also complies with the WordPress philosophy: [decisions, not options](https://wordpress.org/about/philosophy/#decisions).

File: packages/language-picker/src/language-picker.tsx
Comment (sirreal) [reply packages/language-picker/src/language-picker.tsx]: I agree the name is a bit misleading to use `translate` or even `__` given the different contexts `i18n-calypso`, `@wordpress/i18n` and `@automattic/react-i18n`. It may be best to name it something generic like `translationFunction`. What we're describing is the shape of the `name` function in the `LanguageGroup` type, but folks will be free to name it whatever they want:

```ts
const lg: LanguageGroup = {
  //      The name of the translation function passed to my name function
  //         V
  name( translateMyName ) {
    return translateMyName("name");
  }
};
```

Regarding the type, it's probably more appropriate to use `I18nReact["__"]` from `@automattic/react-i18n` as the type of the function because that's ultimately where the function will come from. It's passing the `@wordpress/i18n` `__` type through so it's not too important. `@wordpress/i18n` `__` is unfortunately generic right now (`Function` type) but something that can be improved Gutenberg-side.

Comment (tyxla) [reply packages/language-picker/src/language-picker.tsx]: > is it safe to import __ directly from @wordpress/i18n as long as it is invoked as late as possible, as you said (like during rendering)?

I'd say it's safe to do that. That's why we wrap them in functions, so invoking them late is a good warranty that it'll be done after the locale data has already loaded.

Note: I'm not saying it's a perfect solution. We've been limited in Calypso by that for a while now, too, and we're likely to stay with that limitation in the future.

That being said, I think it's an acceptable trade-off.

File: packages/language-picker/src/language-picker.tsx
Comment (saramarcondes) [reply packages/language-picker/src/language-picker.tsx:95]: Thanks again!

I'll update the README as you suggested.

One thing though, the package no longer uses `react-i18n`, only `wordpress/i18n`. Do you feel that it _should_ be using `react-i18n`? (Admittedly I only changed it because using `react-i18n` was breaking the storybook in a mysterious way)

File: packages/language-picker/src/language-picker.tsx
Comment (sirreal) [reply packages/language-picker/src/language-picker.tsx]: > `@wordpress/i18n` `__` is unfortunately generic right now (`Function` type) but something that can be improved Gutenberg-side.

PR to improve function types: https://github.com/WordPress/gutenberg/pull/26171

File: packages/language-picker/src/language-picker.tsx
Comment (tyxla) [reply packages/language-picker/src/language-picker.tsx:89]: I wouldn't say it's a hard requirement as to which one to use. `react-i18n` is handy, because it adds some syntactic features on top of `@wordpress/i18n`, but since it is generally a wrapper around `@wordpress/i18n`, it doesn't really matter. It basically boils down to:

* Use `react-i18n` if you need to use the HoC and/or the React hook.
* Use `@wordpress/i18n` if you don't need to use any of the above.

Theoretically, it can turn out to be a blocker if `react-i18n` uses an older or newer version of `@wordpress/i18n`, but this shouldn't be a difficult problem to solve, as we have full control over `reach-i18n`, especially if it happens in Calypso.

File: package.json
Comment (tyxla) [reply package.json:157]: FWIW it works well for me now.

File: packages/language-picker/src/style.scss
Comment (tyxla) [inline packages/language-picker/src/style.scss:5]: Same as #46424 - do you think it makes sense to have a common wrapping class for all styles in this package, to make sure all styles are properly encapsulated and have a smaller chance to bleed to outside elements?

File: packages/language-picker/src/style.scss
Comment (tyxla) [inline packages/language-picker/src/style.scss:9]: Nit: I wonder if this `span` can have its own `className` that will always be there; so we could target it by class name and not by element name.

File: packages/language-picker/src/language-picker.tsx
Comment (tyxla) [inline packages/language-picker/src/language-picker.tsx:19]: Unfortunately, we're going to hit into an issue with this now:



Because the existing `LANGUAGE_GROUPS` constant expects that their name will get a translation function as an argument, and we're no longer passing it here, it's going to fail.

We should either:

* Address this in the constant definition, by using `__()` there directly, vs expecting a translation function. Can you think of any downside to this approach, @Automattic/i18n @yuliyan?
* Create a wrapper constant that uses `LANGUAGE_GROUPS` but predefines the `name` by passing `__()` as an argument to the function.

I personally prefer the first approach - ideally, it shouldn't be too hard to replace the instances of `languageGroup.name( translate )` to `languageGroup.name( `__` ) - I expect those to be only in the language picker component anyway.

**Review by tyxla — DISMISSED** [review]
This is looking great overall!

There's one blocker here: the language groups no longer work, because they expect us to pass a translate function as an argument to the language group name function, but we no longer do that.

File: packages/language-picker/src/language-picker.tsx
Comment (saramarcondes) [reply packages/language-picker/src/language-picker.tsx:19]: That makes sense (I thought I had fixed this but I guess I got the Storybook example and the live constants confused 😖)

Something I'm still confused about is whether we need to include the text domain global variable as described by @alshakero in the recent P2 here: pbAok1-1Ao-p2

If we do need to include the `__i18n_text_domain__` variable in the component (both here and in search) then do we need to include it in these constants as well?

Comment (saramarcondes) [reply packages/language-picker/src/language-picker.tsx:19]: @tyxla After making this change actually there is a new but subtle bug. When selecting English from the langauge picker in Gutenboarding (accessed via `/new?flags=gutenboarding/language-picker`) it does not change the region languages to English (whereas for other languages it does work and in production it indeed will change to English when English is selected).

To reproduce:
1. Go to `/new?flags=gutenboarding/language-picker` on this branch
2. Click the language picker from the upper right hand corner
3. Select a languag other than English. Reopen the language picker and note that the regions are now listed in the langauge you selected.
4. Now select English. Reopen the language picker and note that the regions are still listed in the previous language you selected.
5. Repeat this several times going between languages that are not English and note that the regions change every time to the newly selected language _except for when you choose English_.

In production selecting English will behave the same as the rest of the languages.

Comment (simison) [reply packages/language-picker/src/language-picker.tsx:19]: > Something I'm still confused about is whether we need to include the text domain global variable as described by alshakero in the recent P2 here: pbAok1-1Ao-p2

@Automattic/ganon are exploring these questions currently.

Comment (ramonjd) [reply packages/language-picker/src/language-picker.tsx:19]: > @Automattic/ganon are exploring these questions currently.

Latest WIP from @p-jackson here https://github.com/Automattic/wp-calypso/pull/46557

**Comment by saramarcondes** [comment]
@tyxla I've pushed an update to this branch that fixes the translation issues I outlined previously. It makes some rather big changes that I'm not super comfortable with. I'd love your input. I was not able to get it working the way we had it before. The way that we're doing things now will require us to change the Calypso implementation to use `react-i18n` but I think that's okay given we want to move in the direction of using `@wordpress` packages anyway.

I also removed storybook from this package because `react-i18n` is incompatible with it. The story for the component also had very limited usefulness I think, with difficulty we could have made it more useful but not without making some big changes to how we manage and import languages.

In any case. I'm hoping that my changes regarding translations make sense. See if you can find a different way to address the bug if you'd like, or if you can find the root cause for English not getting activated for the component's `__`.

**Comment by tyxla** [comment]
> @tyxla I've pushed an update to this branch that fixes the translation issues I outlined previously. It makes some rather big changes that I'm not super comfortable with. I'd love your input. I was not able to get it working the way we had it before. The way that we're doing things now will require us to change the Calypso implementation to use `react-i18n` but I think that's okay given we want to move in the direction of using `@wordpress` packages anyway.
> 
> I also removed storybook from this package because `react-i18n` is incompatible with it. The story for the component also had very limited usefulness I think, with difficulty, we could have made it more useful but not without making some big changes to how we manage and import languages.
> 
> In any case. I'm hoping that my changes regarding translations make sense. See if you can find a different way to address the bug if you'd like, or if you can find the root cause for English not getting activated for the component's `__`.

I've taken a look at the approach and concluded there isn't an easy way around it right now. If we fall back to using `__()` from `@wordpress/i18n`, in this instance we'll end up in a case where the components that use it aren't subscribed to changes in the locale data. Even if we force a re-render, the `@wordpress/i18n` instance will use the old locale data. This is why we have `react-i18n` in the first place, after all. But the approach there works only for components, and not for static functions that aren't React components.

So to work around it, we should pass a `__()` from `react-i18n` to the constants. That's what you did essentially, and after weighing the trade-offs, I think it's one that we can live with. I actually like the approach, because it makes it more clear why we're doing it, and somewhat isolates the root of the problem.

That being said, feel free to add some documentation in the code to make sure that what we're doing is clear to our future selves. 

I've dismissed my previous review, and from my perspective, this is good to go. Ideally, I'd like to get a ✅  from @Automattic/ganon before merging.

Nice work and thanks for the hard work here 👏

**Comment by jsnajdr** [comment]
> The way that we're doing things now will require us to change the Calypso implementation to use `react-i18n` but I think that's okay given we want to move in the direction of using `@wordpress` packages anyway.

Using the `react-i18n` package is fine and even desired. The plan has always been to develop it in Calypso and then contribute back to `@wordpress` after it's battle-tested.

The `react-i18n` approach is more flexible and more in line with React philosophy than the global singletons and global `__()` functions in the base `@wordpress/i18n` package.

**Comment by tyxla** [comment]
> > The way that we're doing things now will require us to change the Calypso implementation to use `react-i18n` but I think that's okay given we want to move in the direction of using `@wordpress` packages anyway.
> 
> Using the `react-i18n` package is fine and even desired. The plan has always been to develop it in Calypso and then contribute back to `@wordpress` after it's battle-tested.
> 
> The `react-i18n` approach is more flexible and more in line with React philosophy than the global singletons and global `__()` functions in the base `@wordpress/i18n` package.

I was talking about a different trade-off here (passing `__()` down vs being able to directly use it), but just to confirm that I completely agree with what Jarda said.

**Comment by ramonjd** [comment]
> I've dismissed my previous review, and from my perspective, this is good to go. Ideally, I'd like to get a ✅ from @Automattic/ganon before merging.

I don't have anything to add. Maybe @andrewserong @p-jackson or @roo2 have. 

That said, we'll be testing integration as part of the [i18n Gutenboarding effort](https://github.com/Automattic/wp-calypso/issues/37665), so we'll report anything we see then.

Thanks to everyone for their work on this!

File: packages/language-picker/src/language-picker.tsx
Comment (p-jackson) [inline packages/language-picker/src/language-picker.tsx:49]: Optional: you can fix the eslint warning about defining module boundaries by using the `React.FC` type.

```suggestion
const LanguagePicker: React.FC< Props > = ( { onSelectLanguage, languages, languageGroups } ) => {
```

File: packages/language-picker/src/constants.ts
Comment (p-jackson) [inline packages/language-picker/src/constants.ts:5]: Minor nit pick: using [`import type`](https://www.typescriptlang.org/docs/handbook/modules.html#importing-types) makes it easier for tooling to optimise the case where a package is only geing imported for a type.

```suggestion
import type { LanguageGroup } from './language-picker';
import type { I18nReact } from '@automattic/react-i18n';
```

Not a big deal, it just stood out to me because I thought we had an eslint warning about this. Eslint appears to be working in this new package though so I must be mistaken.

**Review by p-jackson — APPROVED** [review]
Tested and worked as expected. LGTM!

The `__i18n_text_domain__` constant is only really needed if we intend to include `@automattic/language-picker` in a WordPress plugin, e.g. the Editing Toolkit. Otherwise leaving it out is fine and it'll default to using the default text domain. This is what Calypso and Gutenboarding use, so it'll work fine.
You've probably seen it, but `__i18n_text_domain__` is defined [here for Calypso](https://github.com/Automattic/wp-calypso/blob/d1a33514c9b97c6d31fa3ddafa9574e8d1fe6510/client/webpack.config.js#L284) and [here for the Editing Toolkit](https://github.com/Automattic/wp-calypso/blob/d1a33514c9b97c6d31fa3ddafa9574e8d1fe6510/apps/editing-toolkit/webpack.config.js#L79). So this package is usable in either of those contexts now. If you try and install this package in a context that doesn't define `__i18n_text_domain__` then this package will crash. But that's fine for now right?

I left a couple of nit picks but they're optional really, I think this is good to go as is, we can iterate on i18n stuff as we go IMO.

**Comment by tyxla** [comment]
> Thanks to everyone for their work on this!

Indeed, high five! Great team effort, folks 🙌

**Comment by saramarcondes** [comment]
Thanks for the feedback @p-jackson. I'll have another follow-up PR to implement this language picker into Calypso and will address the issues you've pointed out there, if that's okay with you! Otherwise I can open a fast-follow PR to specifically address these if you think they should be taken care of sooner.

**Comment by p-jackson** [comment]
> I can open a fast-follow PR to specifically address these if you think they should be taken care of sooner.

@saramarcondes I don't think there's a need, they're the sort of tidy ups that can go whenever

**Pull request merged**
